bindgen/ir/IR.cpp

```diff
@@ -13,48 +14,53 @@ void IR::addFunction(std::string name, std::vector<Parameter *> parameters,
 }
 
 std::shared_ptr<TypeDef> IR::addTypeDef(std::string name,
-                                        std::shared_ptr<Type> type) {
-    typeDefs.push_back(std::make_shared<TypeDef>(std::move(name), type));
+                                        std::shared_ptr<Type> type,
+                                        std::shared_ptr<Location> location) {
+    typeDefs.push_back(
+        std::make_shared<TypeDef>(std::move(name), type, std::move(location)));
     return typeDefs.back();
 }
 
 std::shared_ptr<Type> IR::addEnum(std::string name, const std::string &type,
-                                  std::vector<Enumerator> enumerators) {
+                                  std::vector<Enumerator> enumerators,
+                                  std::shared_ptr<Location> location) {
     std::shared_ptr<Enum> e =
         std::make_shared<Enum>(std::move(name), type, std::move(enumerators));
     enums.push_back(e);
     if (!e->isAnonymous()) {
-        typeDefs.push_back(e->generateTypeDef());
+        typeDefs.push_back(e->generateTypeDef(std::move(location)));
         return typeDefs.back();
     }
     return nullptr;
 }
 
 void IR::addStruct(std::string name, std::vector<Field *> fields,
-                   uint64_t typeSize) {
+                   uint64_t typeSize, std::shared_ptr<Location> location) {
     std::shared_ptr<Struct> s =
         std::make_shared<Struct>(name, std::move(fields), typeSize);
```

Comment:
Again, I'm curious why the location is associated with the generated typedef instead of the "concrete" type.
The latter would require fewer changes, e.g. no need for `TypeDef::setLocation()`.

Reply:
`TypeDef` instance in any case should have `location` because it may reference any type: `typedef int size`.

`Struct`, `Union` and `Enum` are not referenced directly, corresponding `TypeDef` is generated for each instance, so it is okay to use `location` from `TypeDef`.

It is possible to add `location` to them too, but in some sense `TypeDef` replaces actual type, and we store `Struct`, .. only for helper methods.

So when generator checks for unused external types it can go through all typedefs and check their `location`.
If a `TypeDef` is removed and it references a struct, the struct is also removed.

I am not sure if it was right choice, I can try to add `location` to `Struct` etc and see if it will make code more complicated.